### PR TITLE
Add the Gatekeeper constraint status sync controller

### PR DESCRIPTION
**Pull request by mprahl**
This controller is responsible for relaying Gatekeeper constraint audit results as policy status events for constraints associated with a policy.

The controller will only start if Gatekeeper is installed. If Gatekeeper's installation status changes while the controller manager is running, the health endpoint will return as unhealthy to cause Kubernetes to restart the pod to then enable/disable this new controller.

Relates:
https://issues.redhat.com/browse/ACM-3328

### CODE REVIEW

File: main.go
Comment (JustinKuli) [inline main.go:393]: I think this comment is confusing, since we're not conditionally *disabling* it, we're conditionally *enabling* it.
```suggestion
	// Only enable the GatekeeperConstraintReconciler if the Kubernetes version is v1.16 or above
```

File: controllers/gatekeepersync/health.go
Comment (JustinKuli) [inline controllers/gatekeepersync/health.go:39]: If this is the first thing done by the dynamic client (and I think it might be) then I think it could take a long time to do its API discovery... you might already know more about that than me, so correct me if we don't need to be concerned with that here. But if this `List` times out (because of that issue), then the error would bubble up to `main`, and we'd effectively restart the container. But restarting the container will just cause it to have the same problem again, unless I'm missing something.

File: controllers/gatekeepersync/health.go
Comment (JustinKuli) [inline controllers/gatekeepersync/health.go:72]: If the parent context was canceled, should we return a non-nil error? Otherwise it feels like there *could* be some weird situation where it never appears unhealthy, because it keeps hitting this spot?

File: controllers/gatekeepersync/gatekeeper_constraint_sync.go
Comment (JustinKuli) [inline controllers/gatekeepersync/gatekeeper_constraint_sync.go:132]: I believe this only returns a non-nil error if the policyObjID is invalid... it shouldn't be invalid, so it should always be nil, but if it *was* somehow invalid then the error here would cause the policy to be re-queued forever.

**Review by JustinKuli — CHANGES_REQUESTED** [review]
This is only a partial review, but I had some early questions.

**Review by dhaiducek — COMMENTED** [review]
I gave this a first pass, and I don't think I saw anything glaring. I'll look at it again first thing tomorrow when my vviissiioonn isn't blurring. 😉

File: Makefile
Comment (dhaiducek) [reply Makefile:289]: It follows from technical debt in my PR:
> Handling for older Kubernetes/older Gatekeeper (right now, I have a skip-minimum tag that skips the test entirely); Gatekeeper v3.4 added a PodDisruptionBudget that isn't available until Kubernetes v1.21

https://github.com/open-cluster-management-io/governance-policy-framework-addon/pull/33#issuecomment-1450652169

File: test/e2e/case17_gatekeeper_sync_test.go
Comment (mprahl) [reply test/e2e/case17_gatekeeper_sync_test.go:84]: Good catch.

File: controllers/utils/events.go
Comment (mprahl) [reply controllers/utils/events.go:22]: I felt like it was clunky as well. I'm going to embed a struct with a `SendEvent` method that I think will be cleaner.

File: controllers/utils/events.go
Comment (mprahl) [reply controllers/utils/events.go:30]: I think it's good enough for the moment. If you come up with an alternative, please let me know.

File: main.go
Comment (mprahl) [reply main.go:393]: I think I'll remove this semver check and just assume if Gatekeeper is installed, it is working.

File: controllers/gatekeepersync/health.go
Comment (mprahl) [reply controllers/gatekeepersync/health.go:39]: The dynamic client doesn't perform a discovery from what I can tell in the code. It just uses the GVR to construct the Kubernetes API endpoint URL.

File: controllers/gatekeepersync/health.go
Comment (mprahl) [reply controllers/gatekeepersync/health.go:72]: Yes, good point.

File: controllers/gatekeepersync/gatekeeper_constraint_sync.go
Comment (mprahl) [reply controllers/gatekeepersync/gatekeeper_constraint_sync.go:366]: I'll rework this so that if it attempts to send the same status message and it's already in the Policy status, the entry in the lastSentMessages is deleted.

File: test/e2e/case17_gatekeeper_sync_test.go
Comment (mprahl) [reply test/e2e/case17_gatekeeper_sync_test.go:421]: It has a similar message but it truncates it so much that it's hard to debug.

File: controllers/gatekeepersync/policy_predicate.go
Comment (mprahl) [reply controllers/gatekeepersync/policy_predicate.go:16]: It looks like it wasn't needed. I guess it was preemptive on my part.  :laughing:

Comment (JustinKuli) [reply controllers/gatekeepersync/policy_predicate.go]: Maybe we need to re-enable the linter somewhere? Because I would rather not have panics 😄

**Review by JustinKuli — DISMISSED** [review]
/hold

Looks good to me! Thanks for those updates. Leaving it up to @dhaiducek to double-check :)

File: controllers/gatekeepersync/gatekeeper_constraint_sync.go
Comment (dhaiducek) [inline controllers/gatekeepersync/gatekeeper_constraint_sync.go:47]: There isn't concurrency in the other controllers that I can see. Concurrency seems good, but I'm wondering about the effect here.

File: test/e2e/case17_gatekeeper_sync_test.go
Comment (dhaiducek) [inline test/e2e/case17_gatekeeper_sync_test.go:234]: Nit: Following other conventions here. Is there a reason to hardcode this value (as well as the other instances in this suite)?
```suggestion
		}, defaultTimeoutSeconds*4, 1).Should(Succeed())
```

**Review by dhaiducek — APPROVED** [review]
LGTM!
I have a musing about concurrency here where it doesn't seem to be elsewhere and using `defaultTimeoutSeconds` consistently, but they're not anything that should block the PR.

File: controllers/gatekeepersync/gatekeeper_constraint_sync.go
Comment (mprahl) [reply controllers/gatekeepersync/gatekeeper_constraint_sync.go:47]: @dhaiducek you're right. I added concurrency here because by default, the Gatekeeper audits run every minute so if the user has a lot of constraints, I didn't want there to be a delay in getting results.

The way the concurrency works in controller-runtime is that there is only one goroutine processing a particular reconcile request at a time. For example, since we are reconciling `Policy` objects, if there are three reconciles requests for `Policy` named `a` and one reconcile request for `Policy` named `b`, there would just be one goroutine processing `Policy` `a` and one goroutine processing `Policy` `b`.

In other words, it's concurrent across reconciles for different `Policy` objects but reconciles for the same `Policy` will not run concurrently.

File: test/e2e/case17_gatekeeper_sync_test.go
Comment (mprahl) [reply test/e2e/case17_gatekeeper_sync_test.go]: I changed this to use a different constant.

**Pull request merged**
